cpp/src/arrow/array/diff.cc

```diff
@@ -96,45 +99,260 @@ static UnitSlice GetView(const UnionArray& array, int64_t index) {
+  }
+};
+
+class CreateValueComparator {
+ private:
+  std::unique_ptr<ValueComparator> comparator_;
+
+ public:
   template <typename T>
-  Status Visit(const T&) {
+  Status Visit(const T&, const Array& base, const Array& target) {
     using ArrayType = typename TypeTraits<T>::ArrayType;
-    out = [](const Array& base, int64_t base_index, const Array& target,
-             int64_t target_index) {
-      return (GetView(checked_cast<const ArrayType&>(base), base_index) ==
-              GetView(checked_cast<const ArrayType&>(target), target_index));
-    };
+    comparator_ = std::make_unique<DefaultValueComparator<ArrayType>>(
+        checked_cast<const ArrayType&>(base), checked_cast<const ArrayType&>(target));
     return Status::OK();
   }
 
-  Status Visit(const NullType&) { return Status::NotImplemented("null type"); }
+  Status Visit(const NullType&, const Array&, const Array&) {
+    return Status::NotImplemented("null type");
+  }
 
-  Status Visit(const ExtensionType&) { return Status::NotImplemented("extension type"); }
+  Status Visit(const ExtensionType&, const Array&, const Array&) {
+    return Status::NotImplemented("extension type");
+  }
 
-  Status Visit(const DictionaryType&) {
+  Status Visit(const DictionaryType&, const Array& base, const Array& target) {
     return Status::NotImplemented("dictionary type");
   }
 
-  Status Visit(const RunEndEncodedType&) {
-    return Status::NotImplemented("run-end encoded type");
+  Status Visit(const RunEndEncodedType& ree_type, const Array& base,
+               const Array& target) {
+    const auto& base_ree = checked_cast<const RunEndEncodedArray&>(base);
+    const auto& target_ree = checked_cast<const RunEndEncodedArray&>(target);
+    ARROW_ASSIGN_OR_RAISE(
+        auto inner_values_comparator,
+        (*this)(*ree_type.value_type(), *base_ree.values(), *target_ree.values()));
+    auto* ree_value_comparator =
+        ([&ree_type, &base_ree, &target_ree,
+          inner_values_comparator =
+              std::move(inner_values_comparator)]() mutable -> ValueComparator* {
+          const auto run_end_type_id = ree_type.run_end_type()->id();
+          switch (run_end_type_id) {
+            case Type::INT16:
+              return new REEValueComparator<int16_t>(base_ree, target_ree,
+                                                     std::move(inner_values_comparator));
+            case Type::INT32:
+              return new REEValueComparator<int32_t>(base_ree, target_ree,
+                                                     std::move(inner_values_comparator));
+            default:
+              DCHECK_EQ(run_end_type_id, Type::INT64);
+              return new REEValueComparator<int64_t>(base_ree, target_ree,
+                                                     std::move(inner_values_comparator));
+          }
+        }());
+    comparator_ = std::unique_ptr<ValueComparator>{ree_value_comparator};
+    return Status::OK();
```

Comment:
```suggestion
          switch (ree_type.run_end_type()->id()) {
            case Type::INT16:
              comparator_.reset(new REEValueComparator<int16_t>(base_ree, target_ree,
                                                     std::move(inner_values_comparator)));
              return Status::OK();
            case Type::INT32:
              comparator_.reset(new REEValueComparator<int32_t>(base_ree, target_ree,
                                                     std::move(inner_values_comparator)));
              return Status::OK();
            case Type::INT64:
              comparator_.reset(new REEValueComparator<int64_t>(base_ree, target_ree,
                                                     std::move(inner_values_comparator)));
              return Status::OK();
            default:
              Unreachable();
          }
```

Reply:
This exact change cause `clang-tidy` to ask for `make_unique`, but I found a way around that.